sparta/sparta/functional/RegisterDefnsJSON.hpp

```diff
@@ -0,0 +1,235 @@
+#pragma once
+
+#include <rapidjson/document.h>
+#include <rapidjson/filereadstream.h>
+#include <rapidjson/writer.h>
+#include <rapidjson/stringbuffer.h>
+
+namespace sparta
+{
+
+class RegisterDefnsFromJSON
+{
+public:
+    RegisterDefnsFromJSON(const std::vector<std::string>& register_defns_json_filenames)
+    {
+        for (const auto& filename : register_defns_json_filenames) {
+            parse(filename);
+        }
+
+        // Add a definition that indicates the end of the array
+        register_defns_.push_back(RegisterBase::DEFINITION_END);
+    }
+
+    RegisterDefnsFromJSON(const std::string& register_defns_json_filename)
+    {
+        parse(register_defns_json_filename);
+
+        // Add a definition that indicates the end of the array
+        register_defns_.push_back(RegisterBase::DEFINITION_END);
+    }
+
+    RegisterBase::Definition* getAllDefns() {
+        return register_defns_.data();
+    }
+
+private:
+    void parse(const std::string& register_defns_json_filename)
+    {
+        // Read the file into a string
+        std::ifstream ifs(register_defns_json_filename);
+        std::stringstream ss;
+        ss << ifs.rdbuf();
+        std::string json_str = ss.str();
+
+        // Parse the JSON string
+        rapidjson::Document document;
+        document.Parse(json_str.c_str());
+
+        for (auto& item : document.GetArray()) {
+            if (item.HasMember("enabled") && !item["enabled"].GetBool()) {
+                continue;
+            }
+
+            const RegisterBase::ident_type id = item["num"].GetInt();
+
+            cached_strings_.emplace_back(item["name"].GetString());
+            const char *name = cached_strings_.back().raw();
+
+            const RegisterBase::group_num_type group_num = item["group_num"].GetInt();
+            auto iter = group_idx_map_.find(group_num);
+            if (iter == group_idx_map_.end()) {
+                group_idx_map_[group_num] = 0;
+            }
+
+            RegisterBase::group_idx_type group_idx = group_idx_map_[group_num]++;
+            cached_strings_.emplace_back(item["group_name"].GetString());
```

Comment:
Should the `group_name` be optional as well since `group_num` is?

Reply:
Not sure I understand - I am adding "group_name" and "group_num" to every register json. It was my understanding that neither would be optional.